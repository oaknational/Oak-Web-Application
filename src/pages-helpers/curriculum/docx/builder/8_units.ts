import type JSZip from "jszip";

import { cdata, xmlElementToJson } from "../xml";
import { CombinedCurriculumData, Slugs } from "..";
import {
  appendBodyElements,
  insertLinks,
  wrapInBookmarkPoint,
  wrapInLinkTo,
} from "../docx";

import { createCurriculumSlug, createProgrammeSlug } from "./helper";

export default async function generate(
  zip: JSZip,
  { data, slugs }: { data: CombinedCurriculumData; slugs: Slugs },
) {
  const linkDefs: Record<string, string> = {};
  linkDefs["interactiveSequence"] =
    `https://www.thenational.academy/teachers/curriculum/${createCurriculumSlug(
      slugs,
    )}/units`;

  const tier_slug = data.units.find((u) => u.tier_slug)?.tier_slug;
  data.units.forEach((unit, unitIndex) => {
    linkDefs[`unit_${unitIndex}_onlineResources`] =
      `https://www.thenational.academy/teachers/programmes/${createProgrammeSlug(
        unit,
        slugs.examboardSlug,
        tier_slug,
      )}/units/${unit.slug}/lessons`;
  });
  const links = await insertLinks(zip, linkDefs);

  const years = Array.from(
    new Set(data.units.map((unit) => parseInt(unit.year))),
  ).sort((a, b) => a - b);

  const pageXml = `
        <root>
            ${years
              .map((year) => {
                return `
                    <w:p>
                        <w:pPr>
                            <w:pStyle w:val="Heading2"/>
                        </w:pPr>
                        ${wrapInBookmarkPoint(
                          `section_year_${year}`,
                          `
                            <w:r>
                                <w:rPr>
                                    <w:rFonts w:ascii="Arial" w:hAnsi="Arial" w:cs="Arial"/>
                                    <w:color w:val="222222"/>
                                    <w:sz w:val="56"/>
                                    <w:b/>
                                </w:rPr>
                                <w:t>Year ${cdata(year)} units</w:t>
                            </w:r>
                        `,
                        )}
                    </w:p>
                    <w:p>
                        ${wrapInLinkTo(
                          links.interactiveSequence!,
                          `
                            <w:r>
                                <w:rPr>
                                    <w:rFonts w:ascii="Arial" w:hAnsi="Arial" w:cs="Arial"/>
                                    <w:color w:val="222222"/>
                                    <w:sz w:val="24"/>
                                    <w:b/>
                                    <w:u w:val="single"/>
                                </w:rPr>
                                <w:t>View interactive sequence online</w:t>
                            </w:r>
                        `,
                        )}
                    </w:p>
                    <w:p>
                        <w:r>
                            <w:br w:type="page"/>
                        </w:r>
                    </w:p> 
                `;
              })
              .join("")}
            ${data.units
              .map((unit, unitIndex) => {
                return `
                    <w:p>
                        ${wrapInLinkTo(
                          links[`unit_${unitIndex}_onlineResources`]!,
                          `
                            <w:r>
                                <w:t>${unit.title} / Go to unit resources</w:t>
                            </w:r>
                        `,
                        )}
                    </w:p>
                `;
              })
              .join("")}
            ${
              years.length > 0
                ? ""
                : `
                <w:p>
                    <w:r>
                        <w:br w:type="page"/>
                    </w:r>
                </w:p> 
            `
            }          
        </root>
    `;

  await appendBodyElements(zip, xmlElementToJson(pageXml).elements);
}
2.5 - 1.25;
