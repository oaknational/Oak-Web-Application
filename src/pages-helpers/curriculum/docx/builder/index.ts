export { default as frontCover } from "./1_frontCover";
export { default as tableOfContents } from "./2_tableOfContents";
export { default as ourCurriculum } from "./3_ourCurriculum";
export { default as threadsExplainer } from "./4_threadsExplainer";
export { default as subjectExplainer } from "./5_subjectExplainer";
export { default as subjectPrincipals } from "./6_subjectPrincipals";
export { default as ourPartner } from "./7_ourPartner";
export { default as units } from "./8_units";
export { default as threadsOverview } from "./9_threadsOverview";
export { default as threadsDetail } from "./10_threadsDetail";
export { default as backCover } from "./11_backCover";
export { default as pageLayout } from "./12_pageLayout";
