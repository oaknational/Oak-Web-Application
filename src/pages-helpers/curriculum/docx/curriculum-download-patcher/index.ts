import { Element } from "xml-js";

import {
  mapOverElements,
  modifyXmlByRootSelector,
  pipeElementThrough,
  withBlock,
} from "../docx";

import { unitsTablePatch } from "./patches/unitsTable";
import { subjectPatch } from "./patches/subject";
import { tableOfContentsPatch } from "./patches/tableOfContents";
import { subjectExplainerPatch } from "./patches/subjectExplainer";
import { partnerDetailPatch } from "./patches/partnerDetail";
import { partnerNamePatch } from "./patches/partnerName";
import { yearPatch } from "./patches/year";
import { threadsTablePatch } from "./patches/threadsTable";
import { endOfDocumentPatch } from "./patches/endOfDocument";
import { unitTitlePatch } from "./patches/unitTitle";
import { unitNumberPatch } from "./patches/unitNumber";
import { unitLessonsPatch } from "./patches/unitLessons";
import { unitYearPatch } from "./patches/unitYear";
import { unitThreadsPatch } from "./patches/unitThreads";
import { unitPreviousPatch } from "./patches/unitPrevious";
import { unitNextPatch } from "./patches/unitNext";
import { mainThreadsPatch } from "./patches/mainThreads";
import { notUndefined } from "./patches/util";
import { coverPatch } from "./patches/cover";
import { backPatch } from "./patches/back";

import {
  CurriculumOverviewMVData,
  CurriculumUnitsTabData,
} from "@/node-lib/curriculum-api-2023";
import { CurriculumOverviewSanityData } from "@/common-lib/cms-types";

export type CombinedCurriculumData = CurriculumOverviewMVData &
  CurriculumOverviewSanityData &
  CurriculumUnitsTabData;

export default async function CurriculumDownlodsPatch(
  uint8Array: Uint8Array,
  combinedCurriculumData: CombinedCurriculumData,
  examboardSlug: string,
) {
  // NOTE: This is really inefficient at the moment, that's fine for now though
  const moddedFile = await modifyXmlByRootSelector(
    uint8Array,
    "w:document",
    async (doc) => {
      const docMod1 = await mapOverElements(
        doc,
        (el: Element, parent?: Element) => {
          return pipeElementThrough(el, parent, [
            coverPatch(combinedCurriculumData, examboardSlug),
            unitsTablePatch(combinedCurriculumData),
            subjectPatch(combinedCurriculumData),
            tableOfContentsPatch(),
            subjectExplainerPatch(combinedCurriculumData),
            partnerDetailPatch(combinedCurriculumData),
            partnerNamePatch(combinedCurriculumData),
            yearPatch(),
            threadsTablePatch(combinedCurriculumData),
            backPatch(combinedCurriculumData),
            endOfDocumentPatch(),
          ]);
        },
      );

      const docMod2 = await withBlock(
        docMod1!,
        "UNIT_PAGE",
        async (template: Element) => {
          const promises = combinedCurriculumData.units.map((unit, index) => {
            const el = structuredClone(template);
            return mapOverElements(el, (el, parent) => {
              return pipeElementThrough(el, parent, [
                unitTitlePatch(unit),
                unitNumberPatch(unit, index),
                unitLessonsPatch(unit),
                unitYearPatch(unit),
                unitThreadsPatch(unit),
                unitPreviousPatch(unit),
                unitNextPatch(unit),
              ]);
            });
          });
          const elements = await Promise.all(promises);

          return {
            type: "element",
            name: "$FRAGMENT$",
            elements: elements.filter(notUndefined),
          };
        },
      );

      const docMod3 = await withBlock(
        docMod2!,
        "THREAD_PAGE",
        async (el: Element, parent?: Element) => {
          return pipeElementThrough(el, parent, [
            mainThreadsPatch(combinedCurriculumData),
          ]);
        },
      );

      if (!docMod3) {
        throw new Error("Invalid document!");
      }

      return docMod3;
    },
  );

  return moddedFile;
}
