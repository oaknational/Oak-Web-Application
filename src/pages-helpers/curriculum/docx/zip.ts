import NodeBuffer, { Buffer } from "node:buffer";

import JSZip from "jszip";

export async function zipToSimpleObject(zip: JSZip) {
  const output: Record<string, Buffer | string> = {};
  for await (const file of Object.values(zip.files)) {
    const buffer = await file.async("nodebuffer");
    if (NodeBuffer.isUtf8(buffer)) {
      output[file.name] = buffer.toString();
    } else {
      output[file.name] = buffer;
    }
  }
  return output;
}
