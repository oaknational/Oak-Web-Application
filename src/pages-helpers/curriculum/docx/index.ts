import { join } from "path";

import { format } from "date-fns";

import type {
  CurriculumOverviewMVData,
  CurriculumUnitsTabData,
} from "../../../node-lib/curriculum-api-2023";
import type { CurriculumOverviewSanityData } from "../../../common-lib/cms-types";

import * as builder from "./builder";
import {
  cmToEmu,
  cmToTwip,
  createImage,
  generateEmptyDocx,
  insertFooters,
  insertImages,
} from "./docx";

import { Ks4Option } from "@/node-lib/curriculum-api-2023/queries/curriculumPhaseOptions/curriculumPhaseOptions.schema";

export type CombinedCurriculumData = CurriculumUnitsTabData &
  CurriculumOverviewMVData &
  CurriculumOverviewSanityData;

export type Slugs = {
  subjectSlug: string;
  phaseSlug?: string;
  ks4OptionSlug?: string;
  keyStageSlug?: string;
  tierSlug?: string;
  childSubjectSlug?: string;
};

const ENABLE_TIMER = false;
const measure = (key: string, fn: () => void | Promise<void>) => {
  if (!ENABLE_TIMER) {
    const v = fn();
    return v;
  }
  let err: unknown;
  let v;
  console.time(key);
  try {
    v = fn();
  } catch (_err: unknown) {
    err = _err;
  } finally {
    console.timeEnd(key);
  }

  if (err) throw err;
  return v;
};

export default async function docx(
  data: CombinedCurriculumData,
  slugs: Slugs,
  ks4Options: Ks4Option[],
) {
  const zip = await generateEmptyDocx();

  const images = await insertImages(
    zip,
    {
      footerImage: join(
        process.cwd(),
        "src/pages-helpers/curriculum/docx/builder/images/footer-logo.png",
      ),
    },
    "footer-default.xml",
  );

  const footers = await insertFooters(zip, {
    default: `
      <w:p>
        <w:pPr>
            <w:jc w:val="right"/>
        </w:pPr>
        <w:r>
            <w:rPr>
                  <w:rFonts
                    w:ascii="Arial"
                    w:eastAsia="Arial"
                    w:hAnsi="Arial"
                    w:cs="Arial"
                  />
                  <w:rtl w:val="0"/>
            </w:rPr>
            <w:t xml:space="preserve">Exported ${format(
              new Date(),
              "dd MMMM yyyy",
            )}     </w:t>
            <w:tab/>
        </w:r>
        <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
            </w:rPr>
            <w:fldChar w:fldCharType="begin"/>
            <w:instrText xml:space="preserve">PAGE</w:instrText>
            <w:fldChar w:fldCharType="separate"/>
            <w:fldChar w:fldCharType="end"/>
        </w:r>
        <w:r>
            <w:rPr>
                <w:rtl w:val="0"/>
            </w:rPr>
        </w:r>
        <w:r>
            ${createImage(images.footerImage, {
              width: cmToEmu(5.33),
              height: cmToEmu(0.55),
              xPos: cmToEmu(0.001),
              yPos: cmToEmu(0.001),
              xPosAnchor: "margin",
              yPosAnchor: "line",
              isDecorative: true,
            })}
        </w:r>
    </w:p>
    `,
  });

  // Run through the builders
  const runners = {
    frontCover: async () => await builder.frontCover(zip, { data, slugs }),
    frontCoverPageLayout: async () =>
      await builder.pageLayout(zip, {
        margins: {
          top: cmToTwip(1.5),
          right: cmToTwip(1.5),
          bottom: cmToTwip(1.5),
          left: cmToTwip(1.5),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    tableOfContents: async () =>
      await builder.tableOfContents(zip, { data, ks4Options }),
    tableOfContentsPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.25),
          right: cmToTwip(1.25),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.25),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    ourCurriculum: async () => await builder.ourCurriculum(zip),
    threadsExplainer: async () =>
      await builder.threadsExplainer(zip, { slugs, data }),
    subjectExplainer: async () => await builder.subjectExplainer(zip, { data }),
    subjectPrincipals: async () => await builder.subjectPrincipals(),
    ourPartner: async () => await builder.ourPartner(zip, { data }),
    units: async () => await builder.units(zip, { data, slugs, ks4Options }),
    threadsOverview: async () => await builder.threadsOverview(zip, { data }),
    threadsOverviewPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.5),
          right: cmToTwip(1.5),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.5),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    threadsDetail: async () => await builder.threadsDetail(zip, { data }),
    threadsDetailPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.75),
          right: cmToTwip(1.75),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.75),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    backCover: async () => await builder.backCover(zip, { data }),
    backCoverPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        isLast: true,
        margins: {
          top: cmToTwip(1.25),
          right: cmToTwip(1.25),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.25),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
  };

  // Note: Just some code to messure some timings.
  await measure("total", async () => {
    for (const [key, runner] of Object.entries(runners)) {
      await measure(key, async () => {
        await runner();
      });
    }
  });

  return await zip.zipToBuffer();
}
