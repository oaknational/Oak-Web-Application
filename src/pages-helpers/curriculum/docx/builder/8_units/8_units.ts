import { join } from "path";

import { sortBy, uniqBy } from "lodash";

import { cdata, safeXml, xmlElementToJson } from "../../xml";
import { CombinedCurriculumData, Slugs } from "../..";
import {
  insertImages,
  insertLinks,
  wrapInBookmarkPoint,
  wrapInLinkTo,
  createImage,
  cmToEmu,
  appendBodyElements,
  JSZipCached,
} from "../../docx";
import { createCurriculumSlug, generateGridCols } from "../helper";

import { buildUnit } from "./unit_detail";

import {
  CurriculumUnitsFormattedData,
  formatCurriculumUnitsData,
} from "@/pages/teachers/curriculum/[subjectPhaseSlug]/[tab]";
import {
  getSuffixFromFeatures,
  getYearGroupTitle,
} from "@/utils/curriculum/formatting";
import { getUnitFeatures } from "@/utils/curriculum/features";

function generateGroupedUnits(
  data: CurriculumUnitsFormattedData<CombinedCurriculumData["units"][number]>,
) {
  const unitOptions = Object.entries(data.yearData).flatMap(
    ([year, { childSubjects, tiers, units, pathways }]) => {
      let options: {
        year: string;
        tier?: string;
        childSubject?: string;
        pathway?: string;
      }[] = [];

      options.push({
        year,
      });
      if (pathways.length > 0) {
        options = options.flatMap((option) => {
          // TODO: This should be sorted by pathway_display_order, however we need to change the way this is handled.
          return sortBy(pathways, ["pathway_slug"]).map((pathway) => {
            return {
              ...option,
              pathway: pathway.pathway_slug,
            };
          });
        });
      }
      if (childSubjects.length > 0) {
        options = options.flatMap((option) => {
          return childSubjects.map((childSubject) => {
            return {
              ...option,
              childSubject: childSubject.subject_slug,
            };
          });
        });
      }
      if (tiers.length > 0) {
        options = options.flatMap((option) => {
          return tiers.map((tier) => {
            return {
              ...option,
              tier: tier.tier_slug,
            };
          });
        });
      }

      return options.map((option) => {
        return {
          ...option,
          units: units.filter((unit) => {
            if (
              option.tier &&
              unit.tier_slug !== null &&
              unit.tier_slug !== option.tier
            ) {
              return false;
            }
            if (
              option.pathway &&
              unit.pathway_slug !== null &&
              unit.pathway_slug !== option.pathway
            ) {
              return false;
            }
            if (
              option.childSubject &&
              unit.subject_slug !== option.childSubject
            ) {
              return false;
            }
            return true;
          }),
        };
      });
    },
  );

  return unitOptions;
}

export default async function generate(
  zip: JSZipCached,
  { data, slugs }: { data: CombinedCurriculumData; slugs: Slugs },
) {
  const formattedData = formatCurriculumUnitsData(
    data,
  ) as CurriculumUnitsFormattedData<CombinedCurriculumData["units"][number]>;

  const yearXml: string[] = [];
  const groupedUnits = generateGroupedUnits(formattedData);
  for (const { units, year, childSubject, tier, pathway } of groupedUnits) {
    yearXml.push(
      await buildYear(
        zip,
        year,
        formattedData,
        units,
        { childSubject, tier, pathway },
        slugs,
      ),
    );
  }

  const pageXml = safeXml`
    <root>
      ${yearXml.join("")}
      <w:p>
        <w:r>
          <w:br w:type="page" />
        </w:r>
      </w:p>
    </root>
  `;

  await appendBodyElements(zip, xmlElementToJson(pageXml)?.elements);
}

function buildOptions({
  unitOptions,
}: {
  unitOptions: CombinedCurriculumData["units"][number]["unit_options"];
}) {
  if (unitOptions.length > 0) {
    return safeXml`
      <XML_FRAGMENT>
        <w:p>
          <w:pPr>
            <w:spacing w:line="240" w:lineRule="auto" w:after="0" />
          </w:pPr>
          <w:r />
        </w:p>
        <w:p>
          <w:pPr>
            <w:spacing w:line="240" w:lineRule="auto" w:after="0" />
          </w:pPr>
          <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:cs="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
              />
              <w:color w:val="ffffff" />
              <w:shd w:val="clear" w:fill="287c34" />
              <w:rtl w:val="0" />
            </w:rPr>
            <w:t xml:space="preserve">${cdata(
                unitOptions.length,
              )} unit option${cdata(unitOptions.length > 1 ? "s" : "")}</w:t>
          </w:r>
        </w:p>
      </XML_FRAGMENT>
    `;
  }
  return "";
}

function buildEmpty(columnIndex: number) {
  return safeXml`
    <w:tc>
      <w:tcPr>
        <w:tcW w:type="pct" w:w="33.333333333333336%" />
        <w:tcBorders>
          <w:top w:val="single" w:color="FFFFFF" w:sz="48" />
          ${columnIndex > 0
            ? `<w:left w:val="single" w:color="FFFFFF" w:sz="48"/>`
            : ""}
          <w:bottom w:val="single" w:color="FFFFFF" w:sz="48" />
          ${columnIndex < 2
            ? `<w:right w:val="single" w:color="FFFFFF" w:sz="48"/>`
            : ""}
        </w:tcBorders>
        <w:shd w:val="solid" w:color="FFFFFF" w:fill="FFFFFF" />
        <w:tcMar>
          <w:top w:type="dxa" w:w="226" />
          <w:left w:type="dxa" w:w="226" />
          <w:bottom w:type="dxa" w:w="226" />
          <w:right w:type="dxa" w:w="226" />
        </w:tcMar>
      </w:tcPr>
      <w:p />
    </w:tc>
  `;
}

function buildYearColumn({
  index,
  title,
  unitOptions,
}: {
  title: string;
  index: number;
  unitOptions: CombinedCurriculumData["units"][number]["unit_options"];
}) {
  const columnIndex = index % 3;

  return safeXml`
    <w:tc>
      <w:tcPr>
        <w:tcW w:type="pct" w:w="33.333333333333336%" />
        <w:tcBorders>
          <w:top w:val="single" w:color="FFFFFF" w:sz="48" />
          ${columnIndex > 0
            ? `<w:left w:val="single" w:color="FFFFFF" w:sz="48"/>`
            : ""}
          <w:bottom w:val="single" w:color="FFFFFF" w:sz="48" />
          ${columnIndex < 2
            ? `<w:right w:val="single" w:color="FFFFFF" w:sz="48"/>`
            : ""}
        </w:tcBorders>
        <w:shd w:val="solid" w:color="E4F8E0" w:fill="E4F8E0" />
        <w:tcMar>
          <w:top w:type="dxa" w:w="226" />
          <w:left w:type="dxa" w:w="226" />
          <w:bottom w:type="dxa" w:w="226" />
          <w:right w:type="dxa" w:w="226" />
        </w:tcMar>
      </w:tcPr>
      <w:p>
        <w:pPr>
          <w:spacing w:line="240" w:lineRule="auto" w:after="0" />
        </w:pPr>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:cs="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
            />
            <w:b />
            <w:color w:val="222222" />
            <w:sz w:val="44" />
            <w:szCs w:val="44" />
          </w:rPr>
          <w:t xml:space="preserve">${cdata(index + 1)}</w:t>
        </w:r>
      </w:p>
      <w:p>
        <w:pPr>
          <w:spacing w:line="240" w:lineRule="auto" w:after="0" />
        </w:pPr>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:cs="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
            />
            <w:b />
            <w:color w:val="222222" />
            <w:sz w:val="28" />
            <w:szCs w:val="28" />
          </w:rPr>
          <w:t xml:space="preserve">${cdata(title)}</w:t>
        </w:r>
      </w:p>
      ${buildOptions({ unitOptions })}
    </w:tc>
  `;
}

function buildYearRow(children: string) {
  return safeXml`
    <w:tr>
      <w:trPr>
        <w:cantSplit />
      </w:trPr>
      ${children}
    </w:tr>
  `;
}

function removeDups(units: CombinedCurriculumData["units"]) {
  const unitSlugLookup = new Set();
  return units.filter((unit) => {
    const key = unit.slug;
    if (!unitSlugLookup.has(key)) {
      unitSlugLookup.add(key);
      return true;
    }
    return false;
  });
}

type Slug = { childSubject?: string; tier?: string; pathway?: string };

async function buildYear(
  zip: JSZipCached,
  year: string,
  formattedData: CurriculumUnitsFormattedData<
    CombinedCurriculumData["units"][number]
  >,
  unitsInput: CombinedCurriculumData["units"],
  yearSlugs: Slug,
  slugs: Slugs,
) {
  const images = await insertImages(zip, {
    jumpOutArrow: join(
      process.cwd(),
      "src/pages-helpers/curriculum/docx/builder/images/jump-out-arrow-2.png",
    ),
    greenCircle: join(
      process.cwd(),
      "src/pages-helpers/curriculum/docx/builder/images/green-circle.png",
    ),
    underline: join(
      process.cwd(),
      "src/pages-helpers/curriculum/docx/builder/images/underline.png",
    ),
    greenUnderline: join(
      process.cwd(),
      "src/pages-helpers/curriculum/docx/builder/images/green-underline.png",
    ),
  });

  const links = await insertLinks(zip, {
    interactiveSequence: `https://www.thenational.academy/teachers/curriculum/${createCurriculumSlug(
      slugs,
    )}/units`,
  });

  const rows = [];
  const units = removeDups(unitsInput);
  for (let i = 0; i < units.length; i += 3) {
    const unitsColumns = Array(3)
      .fill(true)
      .map((_, colIdx) => {
        const index = i + colIdx;
        const unit = units[index];
        if (unit) {
          return buildYearColumn({
            index: index,
            title: unit.title,
            unitOptions: unit.unit_options,
          });
        } else {
          return buildEmpty(index);
        }
      });
    rows.push(buildYearRow(unitsColumns.join("")));
  }

  const unitsXml: string[] = [];
  const unitUnitsBySlug = uniqBy(units, "slug");

  // FIXME: This is slow
  const tierSlug = units.find((u) => u.tier_slug)?.tier_slug ?? undefined;
  for (let unitIndex = 0; unitIndex < unitUnitsBySlug.length; unitIndex++) {
    const unit = unitUnitsBySlug[unitIndex]!;
    if (unit.unit_options.length > 0) {
      for (
        let unitOptionIndex = 0;
        unitOptionIndex < unit.unit_options.length;
        unitOptionIndex++
      ) {
        const unitOption = unit.unit_options[unitOptionIndex]!;
        unitsXml.push(
          await buildUnit(
            zip,
            unit,
            unitIndex,
            unitOption,
            unitOptionIndex,
            images,
            { ...slugs, tierSlug },
          ),
        );
      }
    } else {
      unitsXml.push(
        await buildUnit(zip, unit, unitIndex, undefined, undefined, images, {
          ...slugs,
          tierSlug,
        }),
      );
    }
  }

  let subjectTierPathwayTitle: undefined | string;

  // For the building this header we can assume all units will contain the same subject/tier/pathway
  const firstUnit = units[0];
  if (firstUnit) {
    if (firstUnit.subject || firstUnit.tier || firstUnit.pathway) {
      subjectTierPathwayTitle = [
        // HERE: Only if child subject is present.
        yearSlugs.childSubject ? firstUnit.subject : undefined,
        firstUnit.tier,
        firstUnit.pathway,
      ]
        .filter(Boolean)
        .join(", ");
    }
  }

  const yearTitleSuffix = [
    getSuffixFromFeatures(getUnitFeatures(firstUnit)),
    "units",
  ]
    .filter(Boolean)
    .join(" ");
  const yearTitle = getYearGroupTitle(
    formattedData.yearData,
    year,
    yearTitleSuffix,
  );

  const isSwimming = formattedData.yearData[year]?.labels.includes("swimming");

  const xml = safeXml`
    <XML_FRAGMENT>
      ${!subjectTierPathwayTitle
        ? ""
        : safeXml`
            <w:p>
              <w:r>
                <w:rPr>
                  <w:rFonts
                    w:ascii="Arial"
                    w:cs="Arial"
                    w:eastAsia="Arial"
                    w:hAnsi="Arial"
                  />
                  <w:color w:val="222222" />
                  <w:sz w:val="28" />
                  <w:szCs w:val="28" />
                </w:rPr>
                <w:t xml:space="preserve">${cdata(
                    subjectTierPathwayTitle,
                  )}</w:t>
              </w:r>
            </w:p>
          `}
      <w:p>
        <w:pPr>
          <w:pStyle w:val="Heading2" />
        </w:pPr>
        ${wrapInBookmarkPoint(
          `section_year_${year}`,
          safeXml`
            <w:r>
              <w:rPr>
                <w:rFonts w:ascii="Arial" w:hAnsi="Arial" w:cs="Arial" />
                <w:b />
                <w:color w:val="222222" />
                <w:sz w:val="56" />
              </w:rPr>
              <w:t>${cdata(yearTitle)}</w:t>
            </w:r>
          `,
        )}
      </w:p>
      ${!isSwimming
        ? ""
        : safeXml`
            <XML_FRAGMENT>
              <w:p>
                <w:r>
                  <w:rPr>
                    <w:rFonts w:ascii="Arial" w:hAnsi="Arial" w:cs="Arial" />
                    <w:color w:val="222222" />
                    <w:sz w:val="24" />
                  </w:rPr>
                  <w:t>
                    ${cdata(
                      "Swimming units should be selected based on the ability and experience of your pupils.",
                    )}
                  </w:t>
                </w:r>
              </w:p>
              <w:p />
            </XML_FRAGMENT>
          `}
      <w:p>
        ${wrapInLinkTo(
          links.interactiveSequence!,
          safeXml`
            <XML_FRAGMENT>
              <w:r>
                <w:rPr>
                  <w:rFonts w:ascii="Arial" w:hAnsi="Arial" w:cs="Arial" />
                  <w:b />
                  <w:color w:val="222222" />
                  <w:sz w:val="24" />
                  <w:u w:val="single" />
                </w:rPr>
                <w:t xml:space="preserve">View interactive sequence online</w:t>
              </w:r>
              <w:r>
                <w:rPr>
                  <w:rFonts w:ascii="Arial" w:hAnsi="Arial" w:cs="Arial" />
                  <w:b />
                  <w:color w:val="222222" />
                  <w:sz w:val="24" />
                  <w:u w:val="none" />
                </w:rPr>
                ${createImage(images.jumpOutArrow, {
                  width: cmToEmu(0.41),
                  height: cmToEmu(0.35),
                  xPosAnchor: "character",
                  yPosAnchor: "line",
                  isDecorative: true,
                })}
              </w:r>
            </XML_FRAGMENT>
          `,
        )}
      </w:p>
      <w:p>
        <w:r>
          <w:rPr>
            <w:rFonts
              w:ascii="Arial"
              w:cs="Arial"
              w:eastAsia="Arial"
              w:hAnsi="Arial"
            />
            <w:color w:val="222222" />
            <w:sz w:val="44" />
            <w:szCs w:val="44" />
          </w:rPr>
          <w:t xml:space="preserve"> </w:t>
        </w:r>
      </w:p>
      <w:tbl>
        <w:tblPr>
          <w:tblW w:type="pct" w:w="100%" />
          <w:tblBorders>
            <w:insideH w:val="single" w:color="auto" w:sz="4" />
            <w:insideV w:val="single" w:color="auto" w:sz="4" />
          </w:tblBorders>
        </w:tblPr>
        <w:tblGrid>${generateGridCols(3)}</w:tblGrid>
        ${rows.join("")}
      </w:tbl>
      <w:p>
        <w:r>
          <w:br w:type="page" />
        </w:r>
      </w:p>
      ${unitsXml}
    </XML_FRAGMENT>
  `;

  return xml;
}
