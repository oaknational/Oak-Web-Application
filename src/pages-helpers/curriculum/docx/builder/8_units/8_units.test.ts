import { generateEmptyDocx } from "../../docx";
import { zipToSimpleObject } from "../../zip";

import generate from "./8_units";
import { data, slugs } from "./8_units.fixture";

describe("8_units", () => {
  it("simple", async () => {
    const zip = await generateEmptyDocx();
    await generate(zip, {
      slugs: slugs,
      data: data,
    });

    expect(zipToSimpleObject(zip.getJsZip())).toMatchSnapshot();
  });
});
