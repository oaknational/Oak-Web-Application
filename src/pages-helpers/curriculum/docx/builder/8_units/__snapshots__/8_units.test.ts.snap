// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`8_units simple 1`] = `Promise {}`;
